frontend/src/metabase/actions/components/ActionViz/Action.tsx

```diff
@@ -84,13 +84,13 @@ export function ActionComponent({
 
   const onSubmit = useCallback(
     (parameterMap: ParametersForActionExecution) => {
+      const action = dashcard.action;
+      const fieldSettings =
+        action?.visualization_settings?.fields ||
+        generateFieldSettingsFromParameters(action?.parameters ?? []);
+
       const params = {
-        ...setNumericValues(
-          dashcardParamValues,
-          generateFieldSettingsFromParameters(
-            dashcard?.action?.parameters ?? [],
-          ),
-        ),
+        ...setNumericValues(dashcardParamValues, fieldSettings),
         ...parameterMap,
       };
```

Comment:
The root cause is that we didn't use the action's form field setting directly, and instead, we tried to generate them out of action parameters. There's a parameter type mismatch between implicit and explicit actions that makes `generateFieldSettingsFromParameters` treat numeric parameters as strings. Although `generateFieldSettingsFromParameters` should never be called for explicit actions what form settings are always defined, but I believe it still might be worth fixing in another PR

Reply:
I noticed that in the only other two places where `generateFieldSettingsFromParameters` is called ([`ActionParmetersInputForm`](https://github.com/metabase/metabase/blob/a90fff13405ffc232ff18e6dce51a8a08245d115/frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.tsx#L97), [`PublicAction`](https://github.com/metabase/metabase/blob/a90fff13405ffc232ff18e6dce51a8a08245d115/frontend/src/metabase/public/containers/PublicAction/PublicAction.tsx#L37)) we are also first checking if `action.visualization_settings.fields` is defined before calling the method as a backup.

To me it seems like a bit of an anti-pattern to manually do this check everywhere we use the function. It could be easy for a developer to forget this check which may cause another bug like this one. Perhaps we should refactor the function or define another function like `getActionFieldSettings(action, backupParams)` that does this check inside of it and use that instead?

Reply:
I agree with @EmmadUsmani that this is not very ergonomic for frontend developers. I think, however, that the better solution is to make implicit and explicit action parameters behave consistently so that type casting can be based on the parameters (not visualization_settings), so we could always just rely on the types provided by the backend for parameters, and leave visualization_settings for more cosmetic things like placeholder text, etc.

Put differently, this code is confusing because our entire parameters situation is confusing. The better fix is to make it less confusing all the way down. 🥴

Reply:
Agreeing re: making parameter behavior consistent. I think consolidating logic is still important, though: right now IMO, it is not clear which form component we should use for actions. There are a couple of options doing slightly different things, but IMO we should just have one

Reply:
I agree that the layers of logic are a little confusing, but not sure we can get away with one component to rule them all without just making it confusing in a different way. We're putting actions forms in a few different places doing a few different things (create action form, run from dashboard, run from model detail). I'd love to see a more elegant structure though!

Reply:
FWIW, I had a draft that looked pretty decent, but it's way outdated with the latest `master` already. I'd love to get back to it once we sort out open issues first